text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+The rules you'd expect apply to `u<N>` and `i<N>`, which is that `u<N>` stores values in the range `0..2.pow(N)`, and `i<N>` stores integers in the range `-2.pow(N - 1)..2.pow(N - 1)`. This means that `u<1>` only holds the values `0` and `1`, and `i<1>` only holds the values `-1` and `0`. The meaning for `u<0>` and `i<0>` is a little less clear, but they both are only allowed to contain the value `0`; the ranges end up being `0..1` and `-0.5..0.5`, which… yeah, zero is the only integer in those ranges, but it still can be confusing.
+
+For now, if you want to ensure that your integers are the right size, you can add a `const { ... }` assertion to your implementations like so:
+
+```rust
+impl<const N: usize> MyDefault for u<N> {
+    fn my_default -> u<N> {
+        const { assert!(N > 1); }
+        1
+    }
+}
+```
+
+This will cause the compiler to fail when `MyDefault` is used for `u<0>` or `u<1>`, since it will force the constant block to be evaluated. Not ideal, but it's the best we've got for now.
+
+## Uncommonly sized integers
+
+One other side effect of having `u<N>` and `i<N>` is that we can represent a lot more types than before. For example, `u<7>` is just a seven-bit integer, which we might use to represent an ASCII character. That said, using fewer bits doesn't necessarily mean you'll use up fewer bits of memory-- for example, `u<7>` by itself still requires a `u<8>` in memory.
+
+Overall, you should expect integers where `N` is not a power of two to take up more size/alignment than their bits might imply. There may be ways of packing the bits together in a way that optimizes the amount of space used, but once you read those values into a `u<N>` or `i<N>`, this is no longer the case.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Basic semantics
+
+The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.
+
+`usize`/`isize` continue to be separate from `u<N>`/`i<N>`, where `N` is the target pointer width, as this also applied to `uN` and `iN`. `bool` will also be made different from `u<1>` for similar reasons, even though casting `false` to `0` and `true` to `1` will remain possible.
+
+`u<N>` are able to store integers in the range `0..2.pow(N)` and `i<N>` are able to store integers in the range `-2.pow(N-1)..2.pow(N-1)`. The cheeky specificity of "integers in the range" ensures that, for `i<0>`, the range `-0.5..0.5` only contains the integer zero; in general, `u<0>` and `i<0>` will need to be special-cased anyway, as they must be ZSTs.
+
+It's always valid to `as`-cast `u<N>` or `i<N>` to `u<M>` or `i<M>`, and the usual sign extension or truncation will occur depending on the bits involved. A few additional casts which are possible:
```

Comment:
> pondering: maybe only allow `as` for the lossless & value-preserving ones? Then the truncating ones can be via a named method that makes that clearer.

you can already do `0x100u16 as u8`, so why not generalize that to all sizes for consistency? restricting `as` to lossless & value-preserving casts is a breaking change.

Reply:
Yeah, the main reason is backwards compatibility here: since we can already do truncating conversions, I don't want to special-case it that you can only do that for explicitly `N = 8, 16, 32, 64, 128` and instead just allowed it everywhere. If we have a change in a future edition where only lossless conversions can be as-casted, then, we can change this behaviour, but I'd rather go with consistency for now.

Reply:
we already have Into for infallible lossless casts.